### PR TITLE
Update exception handling for distributedMV when overwrite

### PR DESCRIPTION
**Pull request by jja725**
### What changes are proposed in this pull request?

Add better cancel handling for distributedMV. Before this change,  we would have the situation that we lose our destination file if we get interrupted soon after we delete the destination file but are not able to write a new file. After this change, we would at least keep our old destination file until the write succeeds. There's still the situation that we fail at the final step for renaming/deleting the destination file so we have both new & old destination file or we only have new destination file with tmp name but we hardly can do anything about it. So when the user retry he would see temporary files.

### Why are the changes needed?

make distributedMV more reliable.

### Does this PR introduce any user facing changes?
After this, when we get exception when writing tmp file, we would try to remove the tmp file(this may fail)
when we get exception when delete old file file, we would left both old file & tmp file in the directory.
when we get exception when rename tmp file, we would left tmp file only in the directory.
tmp file name format would be: destinationUri + String.format(".alluxio.0x%016X.tmp", System.currentTimeMillis())

### CODE REVIEW

File: job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java [hunk 1]
```diff
@@ -165,30 +165,50 @@ private static void migrate(MigrateCommand command,
...
-      } catch (FileAlreadyExistsException e) {
-        if (overwrite) {
-          fileSystem.delete(destinationURI);
-          retry = true;
-        } else {
-          throw e;
+      } while (retry);
+    } catch (Throwable t) {
+      try {
+        if (fileSystem.exists(tmpUri) && !fileSystem.exists(destinationURI)) {
```
Comment (jja725) [inline job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:206]: may need an extra eye on this condition @yuzhu

File: job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java [hunk 2]
```diff
@@ -165,30 +165,52 @@ private static void migrate(MigrateCommand command,
...
         }
-      } catch (FileAlreadyExistsException e) {
-        if (overwrite) {
-          fileSystem.delete(destinationURI);
-          retry = true;
-        } else {
-          throw e;
+      } while (retry);
+    } catch (Throwable t) {
+      // this handles the situation that we get exception after we rename destination file and
```
Comment (yuzhu) [inline job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:203]: I think you mean before rename?

File: job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java [hunk 3]
```diff
@@ -165,30 +165,52 @@ private static void migrate(MigrateCommand command,
...
-           FileOutStream out = fileSystem.createFile(destinationURI, createOptions)) {
-        try {
-          IOUtils.copyLarge(in, out, new byte[8 * Constants.MB]);
-        } catch (Throwable t) {
+    boolean cleanUpRename = false;
+    try {
+      do {
+        retry = false;
+        try (FileInStream in = fileSystem.openFile(new AlluxioURI(source), openFileOptions);
+            FileOutStream out = fileSystem.createFile(destinationURI, createOptions)) {
```
Comment (yuzhu) [inline job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:177]: shouldn't we copy to tmpPath?

**Review by yuzhu — CHANGES_REQUESTED** [review]
could you add some highlevel comments on what this PR does?

File: job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java [hunk 4]
```diff
@@ -180,10 +185,17 @@ private static void migrate(MigrateCommand command,
           }
           throw t;
         }
+        if (cleanUpRename) {
+          fileSystem.delete(tmpUri);
```
Comment (jja725) [inline job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:188]: @yuzhu I update PR description so hopefully it makes sense now. And do you have better idea about how we can better handle file deletion/rename here?

Comment (jja725) [reply job/server/src/main/java/alluxio/job/plan/migrate/MigrateDefinition.java:188]: And want to decide the correct process before I update the unit test.

**Comment by yuzhu** [comment]
@jja725 can you fix the tests and use the assumption that remove and rename fails much less frequently than copy to implement this?

**Comment by yuzhu** [comment]
can you include the tmp files the user expect to see in the PR description in the very rare cases? otherwise LGTM

**Comment by jja725** [comment]
> can you include the tmp files the user expect to see in the PR description in the very rare cases? otherwise LGTM

done

**Review by yuzhu — APPROVED** [review]

**Comment by yuzhu** [comment]
alluxio-bot, merge this please

**Pull request merged**